HW3/Task31/Queue.cs

```diff
@@ -0,0 +1,103 @@
+﻿using System.Collections;
+
+namespace Task31
+{
+    internal class Queue<T> : IQueue<T> where T : struct
+    {
+        public int Capacity
+        {
+            get => _capacity;
+            private set => _capacity = value;
+        }
+
+        private T[]? _arrayOfElements = null!;
+        private int _size;
+        private int _capacity;
+
+        // Constructor to create queue with capacity.
+        public Queue(int capacity = 10)
+        {
+            if (capacity < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(capacity));
+            }
+            _capacity = capacity;
+            _arrayOfElements = new T[_capacity];
+        }
+
+        // Constructor to add element to the queue.
+        public Queue(params T[] arrayOfElements)
+        {
+            ArgumentNullException.ThrowIfNull(arrayOfElements);
+            _capacity = arrayOfElements.Length + 5;
+            _size = arrayOfElements.Length;
+
+            _arrayOfElements = new T[_capacity];
+            Array.Copy(arrayOfElements, _arrayOfElements, _size);
+        }
+
+        // To iterate elements in a loop.
+        public IEnumerator GetEnumerator() => _arrayOfElements!.GetEnumerator();
+
+        // Add a new item to the end of the queue.
+        public void Enqueue(T item)
+        {
+            if (_size == _capacity)
+            {
+                throw new ArgumentOutOfRangeException(nameof(_arrayOfElements));
+            }
+
+            _arrayOfElements![_size] = item;
+            _size++;
+        }
+
+        // Remove first element in queue and return it.
+        public T Dequeue()
+        {
+            if (IsEmpty())
+            {
+                return default(T);
+            }
+
+            // Save first item to variable.
+            var firstItem = _arrayOfElements![0];
+
```

Comment:
do not allocate array here, pure performance )
Memory allocation is valid in constructor, and in Tail() method only. In other locations is forbidden )